doc/generated/pmreorder.1
.\" Automatically generated by Pandoc 2.0.6
.\"
.TH "PMREORDER" "1" "2020-01-16" "PMDK - pmreorder version 1.5" "PMDK Programmer's Manual"
.hy
.\" Copyright 2014-2020, Intel Corporation
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in
.\"       the documentation and/or other materials provided with the
.\"       distribution.
.\"
.\"     * Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived
.\"       from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SH NAME
.PP
\f[B]pmreorder\f[] \- performs a persistent consistency check using a
store reordering mechanism
.SH SYNOPSIS
.IP
.nf
\f[C]
$\ python\ pmreorder\ <options>
\f[]
.fi
.SH DESCRIPTION
.PP
The pmreorder tool is a collection of python scripts designed to parse
and replay operations logged by pmemcheck \- a persistent memory
checking tool.
.PP
Pmreorder performs the store reordering between persistent memory
barriers \- a sequence of flush\-fence operations.
It uses a consistency checking routine provided in the command line
options to check whether files are in a consistent state.
.PP
Considering that logging, replaying and reordering of operations are
very time consuming, it is recommended to use as few stores as possible
in test workloads.
.SH OPTIONS
.PP
\f[C]\-h,\ \-\-help\f[]
.PP
Prints synopsis and list of options.
.PP
\f[C]\-l\ <store_log>,\ \-\-logfile\ <store_log>\f[]
.PP
The pmemcheck log file to process.
.PP
\f[C]\-c\ <prog|lib>,\ \-\-checker\ <prog|lib>\f[]
.PP
Consistency checker type.
.PP
\f[C]\-p\ <path>,\ \-\-path\ <path>\f[]
.PP
Path to the consistency checker.
Checker function has to return 0 for consistent cases and 1 otherwise.
.PP
\f[C]\-n\ <name>,\ \-\-name\ <name>\f[]
.PP
The symbol name of the consistency checking function in the library.
Valid only if the checker type is \f[C]lib\f[].
.PP
\f[C]\-o\ <pmreorder_output>,\ \-\-output\ <pmreorder_output>\f[]
.PP
Set the logger output file.
.PP
\f[C]\-e\ <debug|info|warning|error|critical>,\f[]
.PP
\f[C]\-\-output\-level\ <debug|info|warning|error|critical>\f[]
.PP
Set the output log level.
.PP
\f[C]\-r\ \ <NoReorderNoCheck|\f[]
.PP
\f[C]NoReorderDoCheck|\f[]
.PP
\f[C]ReorderFull|\f[]
.PP
\f[C]ReorderPartial|\f[]
.PP
\f[C]ReorderAccumulative|\f[]
.PP
\f[C]ReorderReverseAccumulative>,\f[]
.PP
\f[C]\-\-default\-engine\ \ <NoReorderNoCheck|\f[]
.PP
\f[C]NoReorderDoCheck|\f[]
.PP
\f[C]ReorderFull|\f[]
.PP
\f[C]ReorderPartial|\f[]
.PP
\f[C]ReorderAccumulative|\f[]
.PP
\f[C]ReorderReverseAccumulative>\f[]
.PP
Set the initial reorder engine.
Default value is \f[C]NoReorderNoCheck\f[].
.PP
\f[C]\-x\ <cli_macros|config_file>,\ \-\-extended\-macros\ <cli_macros|config_file>\f[]
.PP
Assign an engine types to the defined marker.
.PP
\f[C]\-v,\ \-\-version\f[]
.PP
Prints current version of pmreorder.
.SH ENGINES
.PP
By default, the \f[B]NoReorderNoCheck\f[] engine is used, which means
that for each set of stores, the tool will pass\-through all sequences
of stores not reordered and will not run consistency checker on them.
.PP
To enable different types of the reorder engine and begin proper
reordering tests, a number of other engines exist:
.IP \[bu] 2
\f[B]NoReorderDoCheck\f[] \- pass\-through of unchanged operations.
Checks correctness of the stores as they were logged.
Useful for operations that do not require fail safety.
.IP
.nf
\f[C]
Example:
\ \ \ \ \ \ \ \ input:\ (a,\ b,\ c)
\ \ \ \ \ \ \ \ output:\ (a,\ b,\ c)
\f[]
.fi
.IP \[bu] 2
\f[B]ReorderAccumulative\f[] \- checks correctness on a growing subset
of the original sequence.
.IP
.nf
\f[C]
Example:
\ \ \ \ \ \ \ \ input:\ (a,\ b,\ c)
\ \ \ \ \ \ \ \ output:
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ()
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ b,\ c)
\f[]
.fi
.IP \[bu] 2
\f[B]ReorderReverseAccumulative\f[] \- checks correctness on a reverted
growing subset of the original sequence.
.IP
.nf
\f[C]
Example:
\ \ \ \ \ \ \ \ input:\ (a,\ b,\ c)
\ \ \ \ \ \ \ \ output:
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ()
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c,\ b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c,\ b,\ a)
\f[]
.fi
.IP \[bu] 2
\f[B]ReorderPartial\f[] \- checks consistency on 3 randomly selected
sequences from a set of 1000 combinations of the original log, without
repetitions.
.IP
.nf
\f[C]
\ Example:
\ \ \ \ \ \ \ \ \ input:\ (a,\ b,\ c)
\ \ \ \ \ \ \ \ \ output:
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b,\ c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ b,\ c)
\f[]
.fi
.IP \[bu] 2
\f[B]ReorderFull\f[] \- for each set of stores generates and checks
consistency of all possible store permutations.
This might prove to be very computationally expensive for most
workloads.
It can be useful for critical sections of code with limited number of
stores.
.IP
.nf
\f[C]
\ Example:
\ \ \ \ \ \ \ \ input:\ (a,\ b,\ c)
\ \ \ \ \ \ \ \ output:
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ()
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b,\ a)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b,\ c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c,\ a)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c,\ b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ b,\ c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (a,\ c,\ b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b,\ a,\ c)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (b,\ c,\ a)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c,\ a,\ b)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ (c,\ b,\ a)
\f[]
.fi
.PP
When the engine is passed with an \f[C]\-r\f[] option, it will be used
for each logged set of stores.
Additionally, the \f[C]\-x\f[] parameter can be used to switch engines
separately for any marked code sections.
For more details about \f[C]\-x\f[] extended macros functionality see
section INSTRUMENTATION below.
.SH INSTRUMENTATION
.PP
The core of \f[B]pmreorder\f[] is based on user\-provided named markers.
Sections of code can be `marked' depending on their importance, and the
degree of reordering can be customized by the use of various provided
engines.
.PP
For this purpose, Valgrind's pmemcheck tool exposes a generic marker
macro:
.IP \[bu] 2
\f[B]VALGRIND_PMC_EMIT_LOG(value)\f[]
.PP
It emits log to \f[I]store_log\f[] during pmemcheck processing.
\f[I]value\f[] is a user\-defined marker name.
For more details about pmemcheck execution see PMEMCHECK STORE LOG
section below.
.PP
Example:
.IP
.nf
\f[C]
main.c
\&.
\&.
\&.
VALGRIND_PMC_EMIT_LOG("PMREORDER_MEMSET_PERSIST.BEGIN");

pmem_memset_persist(...);

VALGRIND_PMC_EMIT_LOG("PMREORDER_MEMSET_PERSIST.END");
\&.
\&.
\&.
\f[]
.fi
.PP
There are a few rules for macros creation:
.IP \[bu] 2
Valid macro can have any name, but begin and end section have to match
\- they are case sensitive.
.IP \[bu] 2
Macro must have \f[C]\&.BEGIN\f[] or \f[C]\&.END\f[] suffix.
.IP \[bu] 2
Macros can't be crossed.
.PP
Defined markers can be assigned engines types and configured through the
\f[B]pmreorder\f[] tool using the \f[C]\-x\f[] parameter.
.PP
There are two ways to set macro options:
.IP \[bu] 2
Using command line interface in format:
.IP
.nf
\f[C]
PMREORDER_MARKER_NAME1=ReorderName1,PMREORDER_MARKER_NAME2=ReorderName2
\f[]
.fi
.IP \[bu] 2
Using configuration file in .json format:
.IP
.nf
\f[C]
{
\ \ \ \ "PMREORDER_MARKER_NAME1":"ReorderName1",
\ \ \ \ "PMREORDER_MARKER_NAME2":"ReorderName2"
}
\f[]
.fi
.PP
For more details about available engines types, see ENGINES section
above.
.PP
\f[B]libpmemobj\f[](7) and \f[B]libpmem\f[](7) also provide set of
macros that allow to change reordering engine on library or function
level:
.PP
\f[C]<library_name|api_function_name>\f[]
.PP
Example of configuration on function level:
.IP
.nf
\f[C]
{
\ \ \ \ "pmemobj_open":"NoReorderNoCheck",
\ \ \ \ "pmemobj_memcpy_persist":"ReorderPartial"
}
\f[]
.fi
.PP
Example of configuration on library level (affecting all library
functions):
.IP
.nf
\f[C]
{
\ \ \ \ "libpmemobj":"NoReorderNoCheck"
}
\f[]
.fi
.PP
List of marked \f[B]libpmemobj\f[](7) API functions:
.IP
.nf
\f[C]
pmemobj_alloc
pmemobj_cancel
pmemobj_check
pmemobj_close
pmemobj_create
pmemobj_ctl_exec
pmemobj_ctl_set
pmemobj_free
pmemobj_list_insert
pmemobj_list_insert_new
pmemobj_list_move
pmemobj_list_remove
pmemobj_memcpy
pmemobj_memmove
pmemobj_memset
pmemobj_memcpy_persist
pmemobj_memset_persist
pmemobj_open
pmemobj_publish
pmemobj_realloc
pmemobj_reserve
pmemobj_root
pmemobj_root_construct
pmemobj_strdup
pmemobj_tx_abort
pmemobj_tx_add_range
pmemobj_tx_add_range_direct
pmemobj_tx_alloc
pmemobj_tx_commit
pmemobj_tx_free
pmemobj_tx_publish
pmemobj_tx_realloc
pmemobj_tx_strdup
pmemobj_tx_wcsdup
pmemobj_tx_xadd_range
pmemobj_tx_xadd_range_direct
pmemobj_tx_xalloc
pmemobj_tx_zalloc
pmemobj_tx_zrealloc
pmemobj_wcsdup
pmemobj_xalloc
pmemobj_xreserve
pmemobj_zalloc
pmemobj_zrealloc
\f[]
.fi
.PP
List of marked \f[B]libpmem\f[](7) API functions:
.IP
.nf
\f[C]
pmem_memmove
pmem_memcpy
pmem_memset
pmem_memmove_nodrain
pmem_memcpy_nodrain
pmem_memset_nodrain
pmem_memmove_persist
pmem_memcpy_persist
pmem_memset_persist
\f[]
.fi
.SH PMEMCHECK STORE LOG
.PP
To generate \f[I]store_log\f[] for \f[B]pmreorder\f[] run pmemcheck with
additional parameters:
.IP
.nf
\f[C]
valgrind\ \\
\ \ \ \ \-\-tool=pmemcheck\ \\
\ \ \ \ \-q\ \\
\ \ \ \ \-\-log\-stores=yes\ \\
\ \ \ \ \-\-print\-summary=no\ \\
\ \ \ \ \-\-log\-file=store_log.log\ \\
\ \ \ \ \-\-log\-stores\-stacktraces=yes\ \\
\ \ \ \ \-\-log\-stores\-stacktraces\-depth=2\ \\
\ \ \ \ \-\-expect\-fence\-after\-clflush=yes\ \\
\ \ \ \ test_binary\ writer_parameter
\f[]
.fi
.PP
For further details of pmemcheck parameters see pmemcheck
documentation (https://github.com/pmem/valgrind/blob/pmem-3.13/pmemcheck/docs/pmc-manual.xml)
.SH ENVIRONMENT
.PP
By default all logging from PMDK libraries is disabled.
To enable API macros logging set environment variable:
.IP \[bu] 2
\f[B]PMREORDER_EMIT_LOG\f[]=1
.SH EXAMPLE
.IP
.nf
\f[C]
python\ pmreorder.py\ \\
\ \ \ \ \ \ \ \ \-l\ store_log.log\ \\
\ \ \ \ \ \ \ \ \-r\ NoReorderDoCheck\ \\
\ \ \ \ \ \ \ \ \-o\ pmreorder_out.log\ \\
\ \ \ \ \ \ \ \ \-c\ prog\ \\
\ \ \ \ \ \ \ \ \-x\ PMREORDER_MARKER_NAME=ReorderPartial\ \\
\ \ \ \ \ \ \ \ \-p\ checker_binary\ checker_parameter
\f[]
.fi
.PP
Checker binary will be used to run consistency checks on
\[lq]store_log.log\[rq], output of pmemcheck tool.
Any inconsistent stores found during \f[B]pmreorder\f[] analysis will be
logged to \f[C]pmreorder_out.log\f[].
.SH SEE ALSO
.PP
\f[B]<http://pmem.io>\f[]


ChangeLog
Fri Jan 17 2020 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.8-rc1

	This release:
	- Introduces new API in libpmemobj for user-assisted defragmentation
	  (see pmemobj_defrag man page for details).
	- Introduces experimental support for PPC64LE architecture.
	- Introduces new flag in libpmemobj for opting-out of transaction aborts
	  on failure (POBJ_TX_NO_ABORT), along with new variants of existing APIs
	  that didn't accept flags (pmemobj_tx_xfree, pmemobj_tx_xpublish,
	  pmemobj_tx_xlock, pmemobj_tx_xlog_append_buffer, pmemobj_tx_xstrdup,
	  pmemobj_tx_xwcsdup).
	- Moves out libvmem and libvmmalloc to the new repository
	  (https://github.com/pmem/vmem).

	Other changes:
	- obj: introduce new statistics useful for defrag
	- obj: introduce transient/persistent statistics enabling
	- obj: introduce pmemobj_tx_(set/get)_user_data funcs
	- obj: introduce pmemobj_(set/get)_user_data funcs
	- obj: disable workaround for offsetof() since VS 15.5 in public header
	- common: drop support for libndctl < 63 on Linux
	- pool: rename -N --no-exec to -d --dry-run

	Notable bug fixes:
	- obj: fix potential NULL-dereference in ulog_store
	- obj: fix unintended recursive locking during palloc
	- obj: fix lock release order in palloc publish
	- obj: fix transient redo log of more than 64 entries
	- obj: fix capacity ulog calculation
	- obj: fix check of unaligned capacity size
	- rpmem: add a missing case for GPSPM + FLUSH_STRICT
	- pmem: fix pmemcheck support on platforms with eADR
	- pool: fix possible memory leak
	- rpmem: fix possible memory leak in rpmemd_config_read
	- rpmem: fix possible memory leak in rpmemd_log_init
	- rpmem: fix possible use-after-free

Mon Sep 30 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.7

	This release:
	- Introduces new APIs in libpmemobj for managing space used by transactions.
	  (see pmemobj_tx_log_append_buffer man page for details)
	- Introduces new APIs in librpmem, splitting rpmem_persist into rpmem_flush
	  and rpmem_drain, allowing applications to use the flush + drain model
	  already known from libpmem. (libpmemobj does not use this feature yet)
	- Optimizes large libpmemobj transactions by significantly reducing
	  the amount of memory modified at the commit phase.
	- Optimizes tracking of libpmemobj reservations.
	- Adds new flags for libpmemobj's pmemobj_tx_xadd_range[_direct] API:
	  POBJ_XADD_NO_SNAPSHOT and POBJ_XADD_ASSUME_INITIALIZED, allowing
	  applications to optimize how memory is tracked by the library.

	To support some of the above changes the libpmemobj on-media layout had
	to be changed, which means that old pools have to be converted using
	pmdk-convert >= 1.7.

	Other changes:
	- obj: fix merging of ranges when NOFLUSH flag is used (#1100)
	- rpmem: fix closing of ssh connection (#995, #1060)
	- obj: abort transaction on pmemobj_tx_publish failure

	Internal changes:
	- test: fault injection tests for pmemblk, pmemlog, and pmemobj
	- test: improved Python testing framework
	- test: support real pmem in bad blocks tests
	- common: allow not building examples and benchmarks

Tue Aug 27 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.6.1

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains,
	incorporates build fixes for FreeBSD and fixes a number of
	smaller bugs.

	Detailed list of bug fixes:
	- common: (win) fix possible pool file coruption (#972, #715, #603)
	- common: implement correct / robust device_dax_alignment (#1071)
	- obj: fix recycler not locating unused chunks
	- doc: update pmemobj_tx_lock documentation wrt behavior on fail
	- common: fix persistent domain detection (#1093)
	- common: vecq: fix a pointer-to-struct aliasing violation
	  (crash on arm64)
	- common: fix minor issues related to ndctl linking
	- obj: drop recursion from pmemobj_next
	- common: fix bug in badblock file error handling
	- obj: fix handling of malloc failures
	- common: fix handling of malloc failures (ctl)
	- jemalloc: fix build with gcc 9
	- obj: don't overwrite errno when palloc_heap_check_remote fails
	- doc: fix pmreorder emit log macro
	- rpmem: change order of rpmem init (#1043)
	- common: Fix build failure due to unused macro PAGE_SIZE
	- common: support older versions of pkg-config
	- tools: link with release variant of pmemcommon
	- common: add PMDK prefix to local copy of queue.h (#990)
	- rpmem: switch to using an open coded basename (FreeBSD)
	- common: posix_fallocate: guard against integer underflow in
	  check (FreeBSD)
	- test: support Valgrind 3.15
	- test: skip if fi_info is missing
	- test: (win) fix sparsefile error handling
	- test: fix libpmempool_feature tests that match logs
	- test: remove vmem_delete test (#1074)
	- test: adjust matchfiles in vmem_valgrind_region test (#1087)
	- test: remove old log files for windows (#1013)
	- test: remove invalid expect_normal_exit (#1092)
	- test: suppress ld leak (#1098)
	- test: Expose necessary symbols in libvmmalloc_dummy_funcs
	  (FreeBSD)
	- test: fix tests failing because `tput` fails (FreeBSD)
	- test: avoid obj_critnib_mt taking very long on many-core machines
	- test: deal with libndctl's path without build system
	- test: overwrite old log in pmempool_create/TEST14.PS1
	- test: fix match files in tests which use dax devices
	- test: fix match file in rpmem_addr_ext test
	- test: fix pmempool_check test

Wed Aug 28 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.5.2

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains

src/tools/pmreorder/pmreorder.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright 2018-2019, Intel Corporation

import argparse
import statemachine
import opscontext
import consistencycheckwrap
import loggingfacility
import markerparser
import sys
import reorderengines


def main():
    pmreorder_version = "unknown"

    '''
    Argv[1] should be given in order to use -v or --version flag. It is passed
    from the installed script. We check whether argv[1] was given and if it's
    not any of regular parameters we use it as a version of pmreorder and
    remove it from the arguments list.
    '''
    if len(sys.argv) > 1 and sys.argv[1][0] != "-":
        pmreorder_version = sys.argv[1]
        del sys.argv[1]

    # TODO unicode support
    # TODO parameterize reorder engine type
    parser = argparse.ArgumentParser(description="Store reordering tool")
    parser.add_argument("-l", "--logfile",
                        required=True,
                        help="the pmemcheck log file to process")
    parser.add_argument("-c", "--checker",
                        choices=consistencycheckwrap.checkers,
                        default=consistencycheckwrap.checkers[0],
                        help="choose consistency checker type")
    parser.add_argument("-p", "--path",
                        required=True,
                        help="path to the consistency checker and arguments",
                        nargs='+')
    parser.add_argument("-n", "--name",
                        help="consistency check function " +
                        "for the 'lib' checker")
    parser.add_argument("-o", "--output",
                        help="set the logger output file")
    parser.add_argument("-e", "--output-level",
                        choices=loggingfacility.log_levels,
                        help="set the output log level")
    parser.add_argument("-x", "--extended-macros",
                        help="list of pairs MARKER=ENGINE or " +
                        "json config file")
    parser.add_argument("-v", "--version",
                        help="print version of the pmreorder",
                        action="version",
                        version="%(prog)s " + pmreorder_version)
    engines_keys = list(reorderengines.engines.keys())
    parser.add_argument("-r", "--default-engine",
                        help="set default reorder engine " +
                        "default=NoReorderNoChecker",
                        choices=engines_keys,
                        default=engines_keys[0])
    args = parser.parse_args()
    logger = loggingfacility.get_logger(
                                        args.output,
                                        args.output_level)
    checker = consistencycheckwrap.get_checker(
                                               args.checker,
                                               ' '.join(args.path),
                                               args.name)

    markers = markerparser.MarkerParser().get_markers(args.extended_macros)

    # create the script context
    context = opscontext.OpsContext(
                                    args.logfile,
                                    checker,
                                    logger,
                                    args.default_engine,
                                    markers)

    # init and run the state machine
    a = statemachine.StateMachine(statemachine.InitState(context))
    if a.run_all(context.extract_operations()) is False:
        sys.exit(1)


if __name__ == "__main__":
    main()
